ocaml/tests/test_repository.ml

```diff
@@ -67,25 +69,94 @@ let test_introduce_invalid_gpgkey_path () =
   let binary_url = "https://repo.example.com" in
   let source_url = "https://repo-src.example.com" in
   let gpgkey_path_1 = "../some-file" in
+  let origin = `remote in
   Alcotest.check_raises "test_introduce_invalid_gpgkey_path_1"
     Api_errors.(Server_error (Api_errors.invalid_gpgkey_path, [gpgkey_path_1]))
     (fun () ->
       Repository.introduce ~__context ~binary_url ~name_label ~name_description
-        ~source_url ~update:false ~gpgkey_path:gpgkey_path_1
+        ~source_url ~update:false ~gpgkey_path:gpgkey_path_1 ~origin
       |> ignore
     ) ;
   let gpgkey_path_2 = "some.file" in
   Alcotest.check_raises "test_introduce_invalid_gpgkey_path_2"
     Api_errors.(Server_error (Api_errors.invalid_gpgkey_path, [gpgkey_path_2]))
     (fun () ->
       Repository.introduce ~__context ~binary_url ~name_label ~name_description
-        ~source_url ~update:false ~gpgkey_path:gpgkey_path_2
+        ~source_url ~update:false ~gpgkey_path:gpgkey_path_2 ~origin
+      |> ignore
+    )
+
+let test_introduce_duplicate_bundle_repo () =
+  let __context = T.make_test_database () in
+  let name_label = "name" in
+  let name_label_1 = "name1" in
+  let name_description = "description" in
+  let name_description_1 = "description1" in
+  let gpgkey_path = "" in
+  let origin = `bundle in
+  let ref =
+    Repository.introduce ~__context ~name_label ~name_description ~binary_url:""
+      ~source_url:"" ~update:true ~gpgkey_path ~origin
+  in
+  Alcotest.check_raises "test_introduce_duplicate_bundle_repo"
+    Api_errors.(Server_error (repository_already_exists, [Ref.string_of ref]))
+    (fun () ->
+      Repository.introduce ~__context ~name_label:name_label_1
+        ~name_description:name_description_1 ~binary_url:"" ~source_url:""
+        ~update:true ~gpgkey_path ~origin
+      |> ignore
+    )
+
+let test_introduce_bundle_repo_binary_url_not_empty () =
+  let __context = T.make_test_database () in
+  let name_label = "name" in
+  let name_description = "description" in
+  let gpgkey_path = "" in
+  let origin = `bundle in
+  Alcotest.check_raises "test_introduce_bundle_repo_binary_url_not_empty"
+    Api_errors.(Server_error (bundle_repo_url_should_be_empty, []))
+    (fun () ->
+      Repository.introduce ~__context ~name_label ~name_description
+        ~binary_url:"http://test.url" ~source_url:"" ~update:true ~gpgkey_path
```

Comment:
If there is ever a chance that these URLs are used in network request, "example.com` should be used. It is designed for this purpose.